### ISSUE TITLE
Ringbuffer timeouts in certain cases (IDFGH-5624)

### ISSUE THREAD

**Issue by bugadani** [espressif/esp-idf#7344]
I was experiencing some weird issues with regards to the Ringbuffer api, so I wrote a few tests to reproduce my problem. Both of the following tests fail on the "Timeout" assert. The issue is sensitive to the buffer size, the data size and I also had to insert data with different sizes.

<details>
  <summary>Original, more complex examples - Click to expand!</summary>

```c
#include "unity.h"

#include "freertos/FreeRTOS.h"
#include "freertos/ringbuf.h"

#include <stdio.h>
#include <string.h>

TEST_CASE("ringbuf xRingbufferSend problem case", "[ringbuf]") {
    RingbufHandle_t ringbuffer = xRingbufferCreate(200, RINGBUF_TYPE_NOSPLIT);
    uint8_t data[9] = {3, 4, 5, 6, 7, 8, 9, 10, 11};

    for (size_t size = 8; size <= sizeof(data); size++) {
        for (size_t i = 0; i < 40; i++) {
            // Make space by removing records
            while (xRingbufferGetCurFreeSize(ringbuffer) < size) {
                void* item = xRingbufferReceive(ringbuffer, NULL, 0);
                if (item == NULL) {
                    TEST_ASSERT(xRingbufferGetCurFreeSize(ringbuffer) ==
                                xRingbufferGetMaxItemSize(ringbuffer));
                    break;
                }
                // Removed record has the expected data
                TEST_ASSERT_EQUAL_MEMORY(item, data, size);
                vRingbufferReturnItem(ringbuffer, item);
            }
            TEST_ASSERT(xRingbufferGetCurFreeSize(ringbuffer) >= size);

            if (xRingbufferSend(ringbuffer, data, size, 0) != pdTRUE) {
                printf("Failed to acquire memory in event buffer. Required: %u Max: %u Free: %u\n",
                       size, xRingbufferGetMaxItemSize(ringbuffer),
                       xRingbufferGetCurFreeSize(ringbuffer));
                // If there is enough space, xRingbufferSend timed out.
                TEST_ASSERT_MESSAGE(xRingbufferGetCurFreeSize(ringbuffer) < size,
                                    "xRingbufferSend Timeout");
            }
        }

        // Flush buffer
        void* item;
        do {
            item = xRingbufferReceive(ringbuffer, NULL, 0);
            if (item) {
                // Record has expected data
                TEST_ASSERT_EQUAL_MEMORY(item, data, size);
                vRingbufferReturnItem(ringbuffer, item);
            }
        } while (item != NULL);
    }

    vRingbufferDelete(ringbuffer);
}

TEST_CASE("ringbuf xRingbufferSendAcquire problem case", "[ringbuf]") {
    RingbufHandle_t ringbuffer = xRingbufferCreate(200, RINGBUF_TYPE_NOSPLIT);
    uint8_t data[9] = {3, 4, 5, 6, 7, 8, 9, 10, 11};

    for (size_t size = 8; size <= sizeof(data); size++) {
        for (size_t i = 0; i < 40; i++) {
            // Make space by removing records
            while (xRingbufferGetCurFreeSize(ringbuffer) < size) {
                void* item = xRingbufferReceive(ringbuffer, NULL, 0);
                if (item == NULL) {
                    TEST_ASSERT(xRingbufferGetCurFreeSize(ringbuffer) ==
                                xRingbufferGetMaxItemSize(ringbuffer));
                    break;
                }
                // Removed record has the expected data
                TEST_ASSERT_EQUAL_MEMORY(item, data, size);
                vRingbufferReturnItem(ringbuffer, item);
            }
            TEST_ASSERT(xRingbufferGetCurFreeSize(ringbuffer) >= size);

            void* pItem;
            if (xRingbufferSendAcquire(ringbuffer, &pItem, size, 0) == pdTRUE) {
                memcpy(pItem, data, size);

                xRingbufferSendComplete(ringbuffer, pItem);
            } else {
                printf("Failed to acquire memory in event buffer. Required: %u Max: %u Free: %u\n",
                       size, xRingbufferGetMaxItemSize(ringbuffer),
                       xRingbufferGetCurFreeSize(ringbuffer));
                // If there is enough space, xRingbufferSendAcquire timed out.
                TEST_ASSERT_MESSAGE(xRingbufferGetCurFreeSize(ringbuffer) < size,
                                    "xRingbufferSendAcquire Timeout");
            }
        }

        // Flush buffer
        void* item;
        do {
            item = xRingbufferReceive(ringbuffer, NULL, 0);
            if (item) {
                // Record has expected data
                TEST_ASSERT_EQUAL_MEMORY(item, data, size);
                vRingbufferReturnItem(ringbuffer, item);
            }
        } while (item != NULL);
    }

    vRingbufferDelete(ringbuffer);
}
```

</details>

Edit: I've made a simpler test case:
```c
TEST_CASE("simple case", "[ringbuf]") {
    uint8_t data[9] = {3, 4, 5, 6, 7, 8, 9, 10, 11};

    RingbufHandle_t ringbuffer = xRingbufferCreate(80, RINGBUF_TYPE_NOSPLIT);

    TEST_ASSERT(xRingbufferSend(ringbuffer, data, 8, 0));
    TEST_ASSERT(xRingbufferSend(ringbuffer, data, 8, 0));
    TEST_ASSERT(xRingbufferSend(ringbuffer, data, 8, 0));
    TEST_ASSERT(xRingbufferSend(ringbuffer, data, 8, 0));

    void* item;
    item = xRingbufferReceive(ringbuffer, NULL, 0);
    TEST_ASSERT(item != NULL);
    vRingbufferReturnItem(ringbuffer, item);

    item = xRingbufferReceive(ringbuffer, NULL, 0);
    TEST_ASSERT(item != NULL);
    vRingbufferReturnItem(ringbuffer, item);

    item = xRingbufferReceive(ringbuffer, NULL, 0);
    TEST_ASSERT(item != NULL);
    vRingbufferReturnItem(ringbuffer, item);

    TEST_ASSERT(xRingbufferGetCurFreeSize(ringbuffer) >= 9);
    TEST_ASSERT(xRingbufferSend(ringbuffer, data, 10, 0));

    item = xRingbufferReceive(ringbuffer, NULL, 0);
    TEST_ASSERT(item != NULL);
    vRingbufferReturnItem(ringbuffer, item);

    TEST_ASSERT(xRingbufferGetCurFreeSize(ringbuffer) >= 9);
    TEST_ASSERT(xRingbufferSend(ringbuffer, data, 9, 0));

    TEST_ASSERT(xRingbufferGetCurFreeSize(ringbuffer) >= 9);
    TEST_ASSERT(xRingbufferSend(ringbuffer, data, 9, 0));

    TEST_ASSERT(xRingbufferGetCurFreeSize(ringbuffer) >= 9);
    TEST_ASSERT(xRingbufferSend(ringbuffer, data, 9, 0));

    vRingbufferDelete(ringbuffer);
}
```

**Comment by AxelLin** [comment]
What is your esp-idf version? (git describe --tags)

**Comment by bugadani** [comment]
It's dirty due to some accidental saves, but `v4.2`

**Comment by AxelLin** [comment]
```
TEST_ASSERT(xRingbufferSend(ringbuffer, data, 10, 0));
```
You have uint8_t data[9], but pass 10 in above code.
Isn't this overflow?

**Comment by bugadani** [comment]
It's a mistake on my part but doesn't matter. I'll fix the comment

**Comment by Dazza0** [comment]
@bugadani I've confirmed that this indeed a bug. The source of the issue is in `xRingbufferReceive()`. The internal implementation (i.e., `prvGetItemDefault()`) forgets to wrap the read pointer around a dummy data block. This results in the free pointer also being corrupted when the item is returned.

I'll push a fix for this issue shortly. For now, to work around this issue you can replace the following lines of `prvGetItemDefault()` in `ringbuf.c` as such:

From...

```c
     //Check if pucRead requires wrap around
     if ((pxRingbuffer->pucTail - pxRingbuffer->pucRead) < rbHEADER_SIZE) {
         pxRingbuffer->pucRead = pxRingbuffer->pucHead;
     }
```

To 

```c
     //Check if pucRead requires wrap around
     if (((ItemHeader_t *)pxRingbuffer->pucRead)->uxItemFlags & rbITEM_DUMMY_DATA_FLAG ||    //Next item is a dummy
         ((pxRingbuffer->pucTail - pxRingbuffer->pucRead) < rbHEADER_SIZE)) {                //Remaining space cannot fit a header
             pxRingbuffer->pucRead = pxRingbuffer->pucHead;  //Wrap pucREAD around
         }
```

**Comment by bugadani** [comment]
Good catch, thanks!

**Comment by AxelLin** [comment]
> ```c
>      //Check if pucRead requires wrap around
>      if (((ItemHeader_t *)pxRingbuffer->pucRead)->uxItemFlags & rbITEM_DUMMY_DATA_FLAG ||    //Next item is a dummy
>          ((pxRingbuffer->pucTail - pxRingbuffer->pucRead) < rbHEADER_SIZE)) {                //Remaining space cannot fit a header
>              pxRingbuffer->pucRead = pxRingbuffer->pucHead;  //Wrap pucREAD around
>          }
> ```

@Dazza0 

Test on v4.4-dev-3235-g3e370c429624
If I made above code change, I got below assert:
assert failed: prvReturnItemDefault ringbuf.c:604 ((pxCurHeader->uxItemFlags & ( ( UBaseType_t ) 1 )) == 0)

**Comment by AxelLin** [comment]
@Dazza0 
Do you have fix for above assert issue?

**Comment by AxelLin** [comment]
@Dazza0 

I test it using pppos client, the uart driver uses the ringbuffer and I *always* hit below assertion if I change the code as you suggested.
**assert failed: prvReturnItemDefault ringbuf.c:604 ((pxCurHeader->uxItemFlags & ( ( UBaseType_t ) 1 )) == 0)**

ps. It works if I don't change the code as you suggested.

**Comment by 0xjakob** [comment]
Hi @AxelLin, thanks for the feedback. We are still working on this issue at the moment and trying to figure out what the problem with the workaround is.

**Comment by sudeep-mohanty** [comment]
Hi @AxelLin , @bugadani ,
Thanks for reporting this issue and for your feedback.
We've narrowed down the bug to an incorrect use of the comparison operation used in the function `prvGetCurMaxSizeNoSplit()` to get the free size available in a no-spilt buffer. The comparison happens between a signed (`xFreeSize`) and an unsigned (`pxRingbuffer->xMaxItemSize`) operand. When the free size available is less than the no-spilt item header size, `xFreeSize` is a negative value. But the comparison operation treats it as a large unsigned value as is the nature of binary operators, therefore resulting in a free size equal to `xMaxItemSize` when in reality it should be 0.

I'll make a fix for this. In the interim, you could test your code by making the following changes in `prvGetCurMaxSizeNoSplit()` in `components/esp_ringbuf/ringbuf.c` file:

From:

```
    //No-split ring buffer items need space for a header
    xFreeSize -= rbHEADER_SIZE;
    //Limit free size to be within bounds
    if (xFreeSize > pxRingbuffer->xMaxItemSize) {
        xFreeSize = pxRingbuffer->xMaxItemSize;
    } else if (xFreeSize < 0) {
        //Occurs when free space is less than header size
        xFreeSize = 0;
    }
```

To

```
    //No-split ring buffer items need space for a header
    xFreeSize -= rbHEADER_SIZE;
    if (xFreeSize < 0) {
        //Occurs when free space is less than header size
        xFreeSize = 0;
    } else if (xFreeSize > pxRingbuffer->xMaxItemSize) {
        //Limit free size to be within bounds
        xFreeSize = pxRingbuffer->xMaxItemSize;
    }
```

**Comment by sudeep-mohanty** [comment]
Hi @AxelLin,
My bad. I missed mentioning that you do not need the below code change -

```
     //Check if pucRead requires wrap around
-    if ((pxRingbuffer->pucTail - pxRingbuffer->pucRead) < rbHEADER_SIZE) {
+    if (((ItemHeader_t *)pxRingbuffer->pucRead)->uxItemFlags & rbITEM_DUMMY_DATA_FLAG ||    //Next item is a dummy
+        ((pxRingbuffer->pucTail - pxRingbuffer->pucRead) < rbHEADER_SIZE)) {                //Remaining space cannot fit a header
```

Please keep the `pucRead` update as before and only update the `prvGetCurMaxSizeNoSplit()` function as mentioned in an earlier comment.
Would you help try this update and let us know if it works for you?

**Comment by AxelLin** [comment]
Hi @sudeep-mohanty 

It works if only update the prvGetCurMaxSizeNoSplit() function.

I have some suggestions:
1. The change is not trivial, please add a comment to the code.
2. I use pppos_client and the uart driver uses esp_ringbuf heavily.
   However, it seems I didn't hit the timeout issue before changing any code.
   So I cannot verify if the fix actually works or not (because it was working in my use case).
   I suggest add a TEST_CASE for this issue to encure it's correct.

**Comment by AxelLin** [comment]
@sudeep-mohanty 
Need to backport this fix to stable branches.

**Comment by AxelLin** [comment]
Just remind that v4.3 branch still does not include this fix.

**Comment by Alvin1Zhang** [comment]
@AxelLin Thanks for the attention, we have back ported this fix to 4.3, but still under internal reviewing, will sync up here once the fix is available on GitHub.
